File: src/lib-common/collate.js
function collate(x)
{
 check(is_arr,x)
 
 function is_delimited(x)
 {
  if(same(x,"_"))
   return false
   
  if(is_punct(x))
   return true
   
  if(is_space(x))
   return true
   
  return false
 }
 
 const a=arr()
 
 for(const v of x)
 {
  if(is_empty(a))
  {
   push(a,v)
   
   continue
  }
  
  const left1=back(a)
  const left2=back(left1)
  const right1=v
  const right2=front(right1)
  
  if(is_delimited(left2))
  {
  }
  else if(is_delimited(right2))
  {
  }
  else
  {
   push(a,right1)
   
   continue
  }

  const s=concat(left1,right1)
  
  drop(a)
  push(a,s)
 }
 
 return join(a," ")
}

File: src/lib-common/collate.cs
fn collate x:arr
 //is delimited

 fn is_delimited x
  if not is_str x
   ret false

  if same x "_"
   ret false

  if is_punct x
   ret true

  if is_space x
   ret true

  ret false
 end

 //main

 let a arr

 for x
  if is_empty a
   push a v

   cont
  end

  let left back a
  let left back left
  let right front v

  if is_delimited left
  elseif is_delimited right
  else
   push a right

   cont
  end

  let s concat left right

  drop a
  push a s
 end

 ret join a " "
end
